src/main/scala/is/hail/methods/LinearMixedRegression.scala

```diff
@@ -55,50 +61,98 @@ object LinearMixedRegression {
 
     val covNames = "intercept" +: covExpr
 
-    val filteredKinshipMatrix = if (kinshipMatrix.sampleIds sameElements completeSamples)
-      kinshipMatrix
-    else {
-      val fkm = kinshipMatrix.filterSamples(completeSamplesSet)
-      if (!(fkm.sampleIds sameElements completeSamples))
-        fatal("Array of sample IDs in assoc_vds and array of sample IDs in kinship_matrix (with both filtered to complete " +
-          "samples in assoc_vds) do not agree. This should not happen when kinship_matrix is computed from a filtered version of assoc_vds.")
-      fkm
-    }
-
     val n = y.length
     val k = cov.cols
     val d = n - k - 1
 
     if (d < 1)
-      fatal(s"$n samples and $k ${ plural(k, "covariate") } including intercept implies $d degrees of freedom.")
+      fatal(s"lmmreg: $n samples and $k ${plural(k, "covariate")} including intercept implies $d degrees of freedom.")
+
+    info(s"lmmreg: running lmmreg on $n samples with $k sample ${plural(k, "covariate")} including intercept...")
+
+    val (fullU, fullS, rank): (DenseMatrix[Double], DenseVector[Double], Int) = similarityMatrix match {
+      case LDMatrix(irm, variants, nSamplesUsed) => {
+        val variantSet = variants.toSet
+        val localMatrix = new DenseMatrix[Double](irm.numRows().toInt,
+          irm.numCols().toInt,
+          irm.toBlockMatrixDense().toLocalMatrix().toArray)
+
+        val eigK = eigSymD(localMatrix)
+        val V = eigK.eigenvectors
 
-    info(s"lmmreg: running lmmreg on $n samples with $k sample ${ plural(k, "covariate") } including intercept...")
+        val c1 = nSamplesUsed.toDouble / variants.length
+        val S = eigK.eigenvalues.map(d => d * c1)
+        val c2 = 1.0 / math.sqrt(variants.length)
+        val sqrtSInv = S.map(d => c2 / math.sqrt(d))
 
-    val K = new DenseMatrix[Double](n, n, filteredKinshipMatrix.matrix.toBlockMatrixDense().toLocalMatrix().toArray)
+        val filteredVDS = assocVds.filterVariants((v, _, _) => variantSet(v))
+        require(filteredVDS.variants.count() == variantSet.size, "Not all variants in LDMatrix present in vds.")
+
+        // FIXME Clean up this ugliness. Unnecessary back and forth from Breeze to Spark.
+        val VS = V(* , ::) :* sqrtSInv
+        val VSSpark = new linalg.DenseMatrix(VS.rows, VS.cols, VS.toArray)
+
+        import is.hail.distributedmatrix.DistributedMatrix.implicits._
+        val dm = DistributedMatrix[BlockMatrix]
+        import dm.ops._
+
+        val sparkGenotypeMatrix = ToNormalizedIndexedRowMatrix(filteredVDS).toBlockMatrixDense().t
+        val sparkU = (sparkGenotypeMatrix * VSSpark).toLocalMatrix()
```

Comment:
So Dan wrote a method for multiplying a block matrix by a local matrix, and that is what I used here. I haven't compared it against IndexedRowMatrix by local matrix yet. Had a feeling that something we wrote would work better than whatever the Spark people wrote, but will verify